source/slang/syntax.cpp

```diff
@@ -1527,6 +1527,61 @@ void Type::accept(IValVisitor* visitor, void* extra)
             && declRef.Equals(otherWitness->declRef);
     }
 
+    RefPtr<Val> DeclaredSubtypeWitness::SubstituteImpl(Substitutions* subst, int * ioDiff)
+    {
+        DeclRef<GenericTypeParamDecl> genParamDeclRef;
```

Comment:
This seems more complicated than it needs to be, and I don’t think it is actually right to have the `return this` at the end.

A `DeclaredSubtypeWitness` contains three things: the sub-type, the super-type, and a decl-ref to the inheritance of generic constraint decl that witnesses the subtype relationship. Each of those three things can be substituted easily in the simple case.

The case you want to check for here is when the *decl ref* of the witness is a constraint decl belonging to the generic being substituted. You don’t want to use the subtype directly, because we will eventually want to support multiple constraints per type, so that you can’t assume a one-to-one relationship between type parameters and witness parameters.

Reply:
The difficulty here is that without relying on the relationship between a type argument and a witness argument in a GenericSubstitution, I cannnot know which witness argument is meant for this DeclaredSubTypeWitness. The witness argument can be a IRWitnessTable, but the IRWitnessTable itself does not say what DeclaredSubTypeWitness it is for, and in fact the same IRWitnessTable might be used for many difference DeclaredSubTypeWitness objects.

Reply:
> The difficulty here is that without relying on the relationship between a type argument and a witness argument in a GenericSubstitution, I cannnot know which witness argument is meant for this DeclaredSubTypeWitness.

That's what I'm trying to get at. `DeclaredSubtypeWitness::declRef` is a reference to the declaration that represents the witness - this is either an `InheritanceDecl` or a `GenericTypeConstraintDecl`. That latter case is the one that matters for substitution - you just want to check if the `GenericTypeConstraintDecl` is a child of the generic being substituted, and then you can compute the parameter index relatively easily.

The key here is: don't think of a `DeclaredSubtypeWitness` primarily as `DeclaredSubtypeWitness(sub: ..., sup: ...)` and instead think of it as `DeclaredSubtypeWitness(declRef:someInheritanceOrConstraintDecl)`. The only reason the `sub` and `sup` fields exist is because I'm expecting that we'll need other cases of subtype witness down the road.

Once you see a declared subtype witness as always just being a wrapper around a `DeclRef` to either an `InheritanceDecl` or a `GenericTypeConstraintDecl` it is hopefully clear how substitution should work.

The `sub` and `sup` fields are actually redundant, in that they can be extracted from the `DeclRef`.

One thing that all of this points out, though, is that we realistically need to introduce a shared base class for both `GenericTypeConstraintDecl` and `InheritanceDecl` so that we can specialize the `DeclRef` in `DeclaredSubtypeWitness` to that new base class.